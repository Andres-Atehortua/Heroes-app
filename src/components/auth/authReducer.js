import { authTypes } from '../types/types';

export const authReducer = (user = {}, action) => {
  switch (action.type) {
    case authTypes.LOGIN:
      return { ...action.payload, logged: true };

    case authTypes.LOGOUT:
      return { logged: false };

    default:
      return user;
  }
};
