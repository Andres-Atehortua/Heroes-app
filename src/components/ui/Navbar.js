import { useContext } from 'react';
import { Link, NavLink, useHistory } from 'react-router-dom';
import { AuthContext } from '../auth/AuthContext';
import { authTypes } from '../types/types';

const Navbar = () => {
  const history = useHistory();
  const { user, dispatch } = useContext(AuthContext);

  const handleLogout = () => {
    dispatch({ type: authTypes.LOGOUT });
    history.replace('/login');
  };

  return (
    <nav className='navbar navbar-expand-sm navbar-dark bg-dark'>
      <Link className='navbar-brand' to='/'>
        Asociaciones
      </Link>

      <div className='navbar-collapse'>
        <div className='navbar-nav'>
          <NavLink
            activeClassName='active'
            className='nav-item nav-link'
            exact
            to='/marvel'
          >
            Marvel
          </NavLink>

          <NavLink
            activeClassName='active'
            className='nav-item nav-link'
            exact
            to='/dc'
          >
            DC
          </NavLink>

          <NavLink
            activeClassName='active'
            className='nav-item nav-link'
            exact
            to='/search'
          >
            Search
          </NavLink>
        </div>
      </div>

      <div className='navbar-collapse collapse w-100 order-3 dual-collapse2'>
        <ul className='navbar-nav ms-auto me-2'>
          <p className='text-info nav-item nav-link mb-0'>{user.name}</p>
          <button onClick={handleLogout} className=' btn nav-item nav-link'>
            Logout
          </button>
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
