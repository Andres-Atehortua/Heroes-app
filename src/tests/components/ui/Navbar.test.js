import { mount } from 'enzyme';
import { MemoryRouter, Router } from 'react-router';
import { AuthContext } from '../../../components/auth/AuthContext';
import { authTypes } from '../../../components/types/types';
import Navbar from '../../../components/ui/Navbar';

describe('Testing component <Navbar />', () => {
  const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
  };

  const contextValue = {
    user: { name: 'andres', logged: true },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter>
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.text-info').text()).toBe(contextValue.user.name);
  });

  test('should logout when press Logout button and use history', () => {
    wrapper.find('button').simulate('click');

    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: authTypes.LOGOUT,
    });
    expect(historyMock.replace).toHaveBeenCalledWith('/login');
  });
});
