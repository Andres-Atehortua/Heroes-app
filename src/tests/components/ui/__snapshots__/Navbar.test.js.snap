// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component <Navbar /> should match snapshot 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
  >
    <ul
      className="navbar-nav ms-auto me-2"
    >
      <p
        className="text-info nav-item nav-link mb-0"
      >
        andres
      </p>
      <button
        className=" btn nav-item nav-link"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
